runtime/interpreter/interpreter.go

```diff
@@ -5544,25 +5544,66 @@ func (interpreter *Interpreter) trackReferencedResourceKindedValue(
 	values[value] = struct{}{}
 }
 
-func (interpreter *Interpreter) updateReferencedResource(
-	currentStorageID atree.StorageID,
-	newStorageID atree.StorageID,
-	updateFunc func(value ReferenceTrackedResourceKindedValue),
-) {
-	values := interpreter.SharedState.referencedResourceKindedValues[currentStorageID]
+func (interpreter *Interpreter) invalidateReferencedResources(value Value, destroyed bool) {
+	// skip non-resource typed values
+	if !value.IsResourceKinded(interpreter) {
+		return
+	}
+
+	var storageID atree.StorageID
+
+	switch value := value.(type) {
+	case *CompositeValue:
+		value.ForEachLoadedField(interpreter, func(_ string, fieldValue Value) {
+			interpreter.invalidateReferencedResources(fieldValue, destroyed)
+		})
+		storageID = value.StorageID()
+	case *DictionaryValue:
+		value.IterateLoaded(interpreter, func(_, value Value) (resume bool) {
+			interpreter.invalidateReferencedResources(value, destroyed)
+			return true
+		})
+		storageID = value.StorageID()
+	case *ArrayValue:
+		value.IterateLoaded(interpreter, func(element Value) (resume bool) {
+			interpreter.invalidateReferencedResources(element, destroyed)
+			return true
+		})
+		storageID = value.StorageID()
+	case *SomeValue:
+		interpreter.invalidateReferencedResources(value.value, destroyed)
+		return
+	default:
+		// skip non-container typed values.
+		return
```

Comment:
Maybe add a test case here, e.g. referencing an `Int`

Reply:
This line would not be reachable, because we already check for `!value.IsResourceKinded(interpreter)` at the start of this function.
Added the `default` case as a backup.